### DISCUSSION TITLE
How to use transactions?

### DISCUSSION THREAD

**Discussion by johnkwaters** [FirelyTeam/firely-net-sdk#2089] [Q&A]
Is this the correct way to use TransactionBuilder to download a patient, add an Encounter, add a Observation and save the results back?
It executes without error, but returns null.

EDIT: I could REALLY use some help with the TransactionBuilder. There is next to ZERO info on it. Can it created patients and other resources that reference it in one go? Or should the create always happen first? If you update an existing patient and also add observations to it - is this considered an Update or a Create? What is the type to send in the search and resource type of execute? The updated patient or the created observation. It's virtually impossible to figure out how to use this class by trial and error.

```
var pi = ResourceIdentity.Build("Patient", "ORG-SM-05306f3b-c7c3-4f9a-a441-87feb2ec6a66");
var pu = await fc.ReadAsync<Patient>(pi);
pu.Address.First().PostalCode = "01008";

var eid = FHIRHelper.CreateId();
var encounter = new Encounter
{
    Id = eid.Value,
    Period = new Period(new FhirDateTime(2022, 5, 15), new FhirDateTime(2022, 5, 15)),
    Status = Encounter.EncounterStatus.Finished,
};
encounter.Identifier.Add(eid);
encounter.Subject = new ResourceReference( $"Patient/{pu.Id}" );

var oid = FHIRHelper.CreateId();
var observation = new Observation
{
    Id = oid.Value,
    Status = ObservationStatus.Final,
    Code = new CodeableConcept(CodeSystemConstants.LOINC, LOINC.BODY_WEIGHT),
    Value = new Quantity((int)(310/2.2), "kg"),
    Encounter = new ResourceReference($"Encounter/{encounter.Id}"),
    Subject = new ResourceReference($"Patient/{pu.Id}")
};
observation.Identifier.Add(oid);

//var updated = await fc.UpdateAsync(pu);

var transaction = new TransactionBuilder(fc.Endpoint)
    .Update(new SearchParams("_id", pu.Id), pu)
    .Create(encounter)
    .Create(observation);

var b = await fc.executeAsync<Bundle>(transaction.ToBundle(), HttpStatusCode.OK);

```

I tried some other variants, creating a patient with an observation in a transaction. I changed the expected status code to 201, and asked for the client to return the Return.Representation.
It seems the patient got created but not the observation.
Can a patient not be created along with resources in the same transaction?

**Comment by yusukemasuda** [comment]
I am struggling with transactional update, too.
I want to update several resource types in one shot with a transaction.

I tried with a code below, but it failed.
```
var bundle = new TransactionBuilder(client.Endpoint, Bundle.BundleType.TransactionResponse)
                .Update(device.Id, device)
                .Update(serviceRequest.Id, serviceRequest)
                .Update(procedure.Id, procedure)
                .ToBundle();

await client.executeAsync<Bundle>(bundle, System.Net.HttpStatusCode.OK);
```

The error was complaining that returned resource is different from ```Bundle``` as I expected.
```'Operation PUT on Device/57249b6d-a6c9-46d6-bcf5-2d8ce1d0c717 expected a body of type Bundle but a Device was returned'```

I also checked firely-dotnet-sdk source, but I wondered that it doesn't properly process multiple resources.
```
    var request = tx.Entry[0];
    // tx (-> ITyped)? -> entryRequest 
    // entry -> ITyped -> tx
    var entryRequest = await request.ToEntryRequestAsync(Settings).ConfigureAwait(false);

    EntryResponse entryResponse = await Requester.ExecuteAsync(entryRequest).ConfigureAwait(false);
```
The line: 966 is retrieve the first entry only, and it throw a request with it, and further it doen't process the second entry or later.
https://github.com/FirelyTeam/firely-net-sdk/blob/a8e21a935258080899d9ccb1f9fb6afa66708724/src/Hl7.Fhir.Core/Rest/BaseFhirClient.cs#L966

My code got a type-mismatch error raised by the code below:
```
    if (!(result is TResource))
    {
        // If this is an operationoutcome, that may still be allright. Keep the OperationOutcome in 
        // the LastResult, and return null as the result. Otherwise, throw.
        if (result is OperationOutcome)
           return null;

        var message = String.Format("Operation {0} on {1} expected a body of type {2} but a {3} was returned", request.Request.Method,
           request.Request.Url, typeof(TResource).Name, result.GetType().Name);

        Enum.TryParse(response.Response.Status, out HttpStatusCode code);
        throw new FhirOperationException(message, code);
    }
```

The questions I want to ask you:
- Does the firely-dotnet-sdk support transactional update?
- Can the second resource or later in bundle be accepted in the method?
- Does the method support multiple types of resource in one bundle?

I believe all of those are supported upon the FHIR specification.
When the Firely supports those, I would appreciate if tell me how to use the SDK.

**Comment by ewoutkramer** [comment]
> There is next to ZERO info on it

Sorry, that is very true. It's been around for ages, certainly before we started documented new features ;-)  I think it would be good to set that straight.
